src/core/meshcop/dataset_manager.cpp

```diff
@@ -332,21 +334,49 @@ void DatasetManager::SendSet(void)
     }
 }
 
-void DatasetManager::HandleCoapResponse(void *               aContext,
-                                        otMessage *          aMessage,
-                                        const otMessageInfo *aMessageInfo,
-                                        Error                aError)
+void DatasetManager::HandleMgmtSetResponse(void *               aContext,
+                                           otMessage *          aMessage,
+                                           const otMessageInfo *aMessageInfo,
+                                           Error                aError)
 {
-    OT_UNUSED_VARIABLE(aMessage);
-    OT_UNUSED_VARIABLE(aMessageInfo);
-    OT_UNUSED_VARIABLE(aError);
-
-    static_cast<DatasetManager *>(aContext)->HandleCoapResponse();
+    static_cast<DatasetManager *>(aContext)->HandleMgmtSetResponse(
+        *static_cast<Coap::Message *>(aMessage), *static_cast<const Ip6::MessageInfo *>(aMessageInfo), aError);
 }
 
-void DatasetManager::HandleCoapResponse(void)
+void DatasetManager::HandleMgmtSetResponse(Coap::Message &aMessage, const Ip6::MessageInfo &aMessageInfo, Error aError)
 {
-    mCoapPending = false;
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    Error    error;
+    StateTlv stateTlv;
+
+    SuccessOrExit(error = aError);
+    VerifyOrExit(Tlv::FindTlv(aMessage, stateTlv) == kErrorNone, error = kErrorFailed);
+
+    switch (stateTlv.GetState())
+    {
+    case StateTlv::kReject:
+        error = kErrorRejected;
+        break;
+    case StateTlv::kAccept:
+        error = kErrorNone;
+        break;
+    default:
+        error = kErrorFailed;
+        break;
+    }
+
+exit:
+    otLogInfoMeshCoP("MGMT_SET finished: %s", ErrorToString(error));
+
+    if (mMgmtSetCallback != nullptr)
+    {
+        mMgmtSetCallback(error, mMgmtSetCallbackContext);
+        mMgmtSetCallback        = nullptr;
+        mMgmtSetCallbackContext = nullptr;
```

Comment:
```suggestion
        auto callback = mMgmtSetCallback;
        auto context = mMgmtSetCallbackContext;
        
        mMgmtSetCallback        = nullptr;
        mMgmtSetCallbackContext = nullptr;
        
        callback(error, context);
```

Should clear callback and context first in case user may start another MGMT_SET request in the callback.